### PR TITLE
New CUDA simulator.

### PR DESCRIPTION
**Pull request by sergeisakov**

### CODE REVIEW

**Review by 95-martin-orion — COMMENTED** [review]
The smaller changes all look reasonable. In lieu of a 10k-line code review, could you summarize the changes made to `simulator_cuda[_kernels].h`?

File: tests/simulator_testfixture.h [hunk 1]
```diff
@@ -1147,60 +1147,63 @@ void TestMultiQubitGates(const Factory& factory) {
...
+  unsigned max_num_qubits = 10 + std::log2(Simulator::SIMDRegisterSize());
 
   StateSpace state_space = factory.CreateStateSpace();
   Simulator simulator = factory.CreateSimulator();
 
-  auto state = state_space.Create(num_qubits);
-
   std::vector<fp_type> matrix;
-  matrix.reserve(1 << (2 * max_gate_qubits + 1));
+  matrix.reserve(1 << (2 * 6 + 1));
```
[the remark this answers is not in the document]
Comment (sergeisakov) [reply tests/simulator_testfixture.h]: Fixed.

**Comment by sergeisakov** [comment]
> The smaller changes all look reasonable. In lieu of a 10k-line code review, could you summarize the changes made to `simulator_cuda[_kernels].h`?

Summary of the changes:

1. The new code is written by hand from scratch and it is much shorter. The old code was generated.

2. The new code uses shared memory. This leads to performance improvements, especially, for 5-qubit gates and for gates with "low" qubits.

3. Controlled gates are faster because the indices are calculated in a slightly different way (in addition to using shared memory).

4. `cudaMemcpyAsync` is used instead of `cudaMemcpy`. This leads to significant performance improvements for small circuits (say, less than 20 qubits).

5. The number of threads is fixed now (64 threads for gates with "high" qubits only and 32 threads for gates with "low" qubits).

**Review by 95-martin-orion — APPROVED** [review]
Does (2) change the optimal gate-fusion size? If so, we should update the docs that mention it.

Otherwise all looks good to me.

**Comment by sergeisakov** [comment]
> Does (2) change the optimal gate-fusion size? If so, we should update the docs that mention it.

Yes, f=5 can be optimal now for some circuits. It depends on the circuit structure.
